### DISCUSSION TITLE
x86_64平台 API：30 调用栈偏移问题

### DISCUSSION THREAD

**Discussion by Mr-JingShi** [hexhacking/xUnwind#9] [Q&A] [answered]
老师你好，x86_64平台 API：30

libsample反汇编
126c:	e8 4f fa ff ff       	callq  cc0 <tgkill@plt>
1271:	bf 04 00 00 00       	mov    $0x4,%edi

CFI方式：
I/xunwind_tag: #01 pc 0000000000001270  /data/app/~~WZ_mVkK4mGpxxUIdnHz4Mg==/io.hexhacking.xunwind.sample-B_YKIacqKu-eUUOxomuhFQ==/lib/x86_64/libsample.so
与实际调用栈偏移4个字节。

EH方式：
I/xunwind_tag: #05 pc 0000000000001271  /data/app/~~WZ_mVkK4mGpxxUIdnHz4Mg==/io.hexhacking.xunwind.sample-B_YKIacqKu-eUUOxomuhFQ==/lib/x86_64/libsample.so
与实际调用栈偏移5个字节。

这是怎么回事呢？
我想做一个自动化的转换工具，该怎么兼容呢？

**Comment by Mr-JingShi** [comment, marked as answer]
> 老师你好，x86_64平台 API：30
> 
> libsample反汇编 126c: e8 4f fa ff ff callq cc0 [tgkill@plt](mailto:tgkill@plt) 1271: bf 04 00 00 00 mov $0x4,%edi
> 
> CFI方式： I/xunwind_tag: #1 pc 0000000000001270 /data/app/~~WZ_mVkK4mGpxxUIdnHz4Mg==/io.hexhacking.xunwind.sample-B_YKIacqKu-eUUOxomuhFQ==/lib/x86_64/libsample.so 与实际调用栈偏移4个字节。
> 
> EH方式： I/xunwind_tag: #5 pc 0000000000001271 /data/app/~~WZ_mVkK4mGpxxUIdnHz4Mg==/io.hexhacking.xunwind.sample-B_YKIacqKu-eUUOxomuhFQ==/lib/x86_64/libsample.so 与实际调用栈偏移5个字节。
> 
> 这是怎么回事呢？ 我想做一个自动化的转换工具，该怎么兼容呢？

```
uint64_t GetPcAdjustment(uint64_t rel_pc, Elf* elf, ArchEnum arch) {
  switch (arch) {
    case ARCH_ARM: {
      if (!elf->valid()) {
        return 2;
      }

      uint64_t load_bias = elf->GetLoadBias();
      if (rel_pc < load_bias) {
        if (rel_pc < 2) {
          return 0;
        }
        return 2;
      }
      uint64_t adjusted_rel_pc = rel_pc - load_bias;
      if (adjusted_rel_pc < 5) {
        if (adjusted_rel_pc < 2) {
          return 0;
        }
        return 2;
      }

      if (adjusted_rel_pc & 1) {
        // This is a thumb instruction, it could be 2 or 4 bytes.
        uint32_t value;
        if (!elf->memory()->ReadFully(adjusted_rel_pc - 5, &value, sizeof(value)) ||
            (value & 0xe000f000) != 0xe000f000) {
          return 2;
        }
      }
      return 4;
    }
  case ARCH_ARM64: {
    if (rel_pc < 4) {
      return 0;
    }
    return 4;
  }
  case ARCH_MIPS:
  case ARCH_MIPS64: {
    if (rel_pc < 8) {
      return 0;
    }
    // For now, just assume no compact branches
    return 8;
  }
  case ARCH_X86:
  case ARCH_X86_64: {
    if (rel_pc == 0) {
      return 0;
    }
    return 1;
  }
  case ARCH_UNKNOWN:
    return 0;
  }
}
```

```
if (adjust_pc) {
  pc_adjustment = GetPcAdjustment(rel_pc, elf, arch_);
} else {
  pc_adjustment = 0;
}
step_pc -= pc_adjustment;
```

和GetPcAdjustment有关系吗？

**Reply by caikelun** [reply]
是的，和PcAdjustment有关，backtrace回溯的其实是LR（返回地址），就是函数返回后继续执行的下一条指令的地址，一般需要把LR减去一个数值。PcAdjustment做的最准确应该是CIF的remote方式，系统的tombstone就是用这个获取backtrace的，你可以验证一下。x86_64我其实没有仔细验证过正确性，你可以研究下。
